core/src/main/scala/kafka/tools/ConsoleConsumer.scala

```diff
@@ -413,13 +411,17 @@ object ConsoleConsumer extends Logging {
       }
     }
 
-    def seek(topic: String, partitionId: Int, offset: Long): Unit = {
+    def seek(topic: String, partitionId: Int, offset: String): Unit = {
       val topicPartition = new TopicPartition(topic, partitionId)
       consumer.assign(Collections.singletonList(topicPartition))
       offset match {
-        case ListOffsetRequest.EARLIEST_TIMESTAMP => consumer.seekToBeginning(Collections.singletonList(topicPartition))
-        case ListOffsetRequest.LATEST_TIMESTAMP => consumer.seekToEnd(Collections.singletonList(topicPartition))
-        case _ => consumer.seek(topicPartition, offset)
+        case EARLIEST_STRING => consumer.seekToBeginning(Collections.singletonList(topicPartition))
+        case LATEST_STRING => consumer.seekToEnd(Collections.singletonList(topicPartition))
+        case _ =>
+          // No need to check if `offset` can convert to a long. We have already done that during the config initialization.
+          val targetOffset = if (offset.toLong < 0)
+            consumer.endOffsets(Collections.singleton(topicPartition)).get(topicPartition) + offset.toLong else offset.toLong
```

Comment:
We need to guard against the case where this sum results in a negative number (to avoid an exception from the tool). In my opinion, in this case, the outcome should be similar (with proper error message) to how the tool responds to a negative offset without this PR.

A unit test to cover this scenario would also be great to have.

Reply:
Great catch! Instead of throwing an error message, what about seeking to the earliest offset if an extremely small negative value is offered, like what `tail -n` did?

Reply:
I thought about that option too. I think that would be fine, but perhaps it would help user get an understanding of the partition offsets in that scenario if a warning message is also reported. Otherwise, there is no way to tell whether the `current offset - N` is a valid offset or not.

I also realized something else: We also need to properly handle the special case when `current offset - N` is a positive number but corresponds to an expired offset. Current PR does not return any messages in this situation (hangs). Similar to above, a warning message and resetting offsets to the earliest available offset could work.

We also need to make sure the output is consistent with partition offsets if the option `num-messages` is also provided. For example is earliest and latest offsets are `10` and `20`, respectively, the combination `--offsets -15 --num-messages 3` should not result in any messages because on messages exist in the offset range `[5, 7]`

Reply:
For the num-messages things, I think it's okay to return messages of offset 10,11,12. What do you think?

Reply:
Hmm. Currently, if the offset parameter points to an expired offset, the console consumer tool gets stuck. So maybe keeping the same behavior and keeping things more consistent would make sense (the current implementation already provides that). Do you agree?